File: src/hilbert/models/paper.py
"""Paper models for Hilbert."""

from datetime import date
from typing import Optional
from pydantic import BaseModel, Field, HttpUrl


class Author(BaseModel):
    """Author of a paper."""

    name: str
    affiliation: Optional[str] = None
    author_id: Optional[str] = None


class Paper(BaseModel):
    """Academic paper from ArXiv or Semantic Scholar."""

    paper_id: str
    title: str
    abstract: str
    authors: list[Author]
    published_date: Optional[date] = None
    url: HttpUrl
    arxiv_id: Optional[str] = None
    doi: Optional[str] = None
    venue: Optional[str] = None
    citation_count: int = 0
    is_open_access: bool = True

    def citation_key(self) -> str:
        """Generate BibTeX citation key."""
        if self.doi:
            return self.doi.replace("/", "-").replace(".", "-")
        if self.arxiv_id:
            return f"arxiv-{self.arxiv_id}"
        return f"paper-{self.paper_id[:8]}"

    def to_bibtex(self) -> str:
        """Convert to BibTeX format."""
        first_author = self.authors[0].name.split()[-1] if self.authors else "Unknown"
        year = self.published_date.year if self.published_date else "n.d."
        key = f"{first_author}{year}"

        entries = [
            f"@article{{{key},",
            f"  author = {{{' and '.join(a.name for a in self.authors)}}},",
            f"  title = {{{self.title}}},",
        ]

        if self.venue:
            entries.append(f"  journal = {{{self.venue}}},")
        if self.published_date:
            entries.append(f"  year = {{{self.published_date.year}}},")
        if self.doi:
            entries.append(f"  doi = {{{self.doi}}},")
        if self.url:
            entries.append(f"  url = {{{self.url}}},")

        entries.append("}")
        return "\n".join(entries)

File: cli/src/core/models/paper.ts
import { z } from "zod";

export const AuthorSchema = z.object({
  name: z.string(),
  affiliation: z.string().optional(),
  author_id: z.string().optional(),
});

export type Author = z.infer<typeof AuthorSchema>;

export const PaperSchema = z.object({
  paper_id: z.string(),
  title: z.string(),
  abstract: z.string(),
  authors: z.array(AuthorSchema),
  published_date: z.string().optional(),
  url: z.string(),
  arxiv_id: z.string().optional(),
  doi: z.string().optional(),
  venue: z.string().optional(),
  citation_count: z.number().default(0),
  is_open_access: z.boolean().default(true),
});

export type Paper = z.infer<typeof PaperSchema>;

export function citationKey(paper: Paper): string {
  if (paper.doi) return paper.doi.replace(/\//g, "-").replace(/\./g, "-");
  if (paper.arxiv_id) return `arxiv-${paper.arxiv_id}`;
  return `paper-${paper.paper_id.slice(0, 8)}`;
}

export function toBibtex(paper: Paper): string {
  const firstAuthor = paper.authors[0]?.name.split(" ").slice(-1)[0] || "Unknown";
  const year = paper.published_date ? new Date(paper.published_date).getFullYear() : "n.d.";
  const key = `${firstAuthor}${year}`;

  const entries = [
    `@article{${key},`,
    `  author = {${paper.authors.map((a) => a.name).join(" and ")}},`,
    `  title = {${paper.title}},`,
  ];

  if (paper.venue) entries.push(`  journal = {${paper.venue}},`);
  if (paper.published_date) {
    entries.push(`  year = {${new Date(paper.published_date).getFullYear()}},`);
  }
  if (paper.doi) entries.push(`  doi = {${paper.doi}},`);
  if (paper.url) entries.push(`  url = {${paper.url}},`);

  entries.push("}");
  return entries.join("\n");
}

export function toPaperText(paper: Paper): string {
  const truncatedAbstract =
    paper.abstract.length > 2000 ? paper.abstract.slice(0, 2000) + "..." : paper.abstract;
  return `${paper.title}\n\n${truncatedAbstract}`;
}